src/main/java/net/neoforged/neoforge/client/event/AddSectionGeometryEvent.java

```diff
@@ -0,0 +1,162 @@
+/**
+ * This event can be used to add static geometry to chunk sections. The event is fired on the main client thread
+ * whenever a section is queued for (re)building. A rebuild can be triggered manually using e.g.
+ * {@link net.minecraft.client.renderer.LevelRenderer#setSectionDirty(int, int, int)}.<br>
+ *
+ * While the event itself is fired on the main client thread, the renderers registered using
+ * {@link net.neoforged.neoforge.client.event.AddSectionGeometryEvent#addRenderer(net.neoforged.neoforge.client.event.AddSectionGeometryEvent.AdditionalSectionRenderer)}
+ * while be executed on the thread performing the rebuild, which will typically <b>not</b> be the main thread.
+ * Therefore, any data from non-thread-safe data structures need to be retrieved during the event handler itself rather
+ * than the renderer. A typical usage would look like
+ * <pre>
+ * {@code
+ * @SubscribeEvent
+ * public static void addChunkGeometry(AddSectionGeometryEvent ev) {
+ *     if (shouldAddGeometryTo(ev.getLevel(), ev.getSectionOrigin())) {
+ *         final var renderingData = getDataOnMainThread(ev.getLevel(), ev.getSectionOrigin());
+ *         ev.addRenderer(context -> renderThreadsafe(renderingData, context));
+ *     }
+ * }
+ * }
+ * </pre>
+ * Note that the renderer is only added if something will actually be rendered in this example. This structure should be
+ * replicated whenever the event is used, to allow for optimizations related to entirely empty sections.
+ *
+ * <p>This event is not {@linkplain ICancellableEvent cancellable}</p>
+ *
+ * <p>This event is fired on the {@linkplain NeoForge#EVENT_BUS main Forge event bus}, only on the
+ * {@linkplain LogicalSide#CLIENT logical client}.</p>
+ */
+public class AddSectionGeometryEvent extends Event {
+    private final List<AdditionalSectionRenderer> additionalRenderers = new ArrayList<>();
+    private final BlockPos sectionOrigin;
+    private final Level level;
+
+    public AddSectionGeometryEvent(BlockPos sectionOrigin, Level level) {
+        this.sectionOrigin = sectionOrigin;
+        this.level = level;
+    }
+
+    /**
+     * Adds a renderer which will add geometry to the chunk section.
+     * @param renderer the renderer to add
+     */
+    public void addRenderer(AdditionalSectionRenderer renderer) {
+        additionalRenderers.add(renderer);
+    }
+
+    /**
+     * @return the list of all added renderers. Do not modify the result.
+     */
+    public List<AdditionalSectionRenderer> getAdditionalRenderers() {
+        return additionalRenderers;
+    }
+
+    /**
+     * @return the origin of the section to add renderers to, i.e. the block with the smallest coordinates contained in
+     * the section.
+     */
+    public BlockPos getSectionOrigin() {
+        return sectionOrigin;
+    }
+
+    /**
+     * @return the level to render in. This can differ from the current client level in case of e.g. guidebooks.
+     */
+    public Level getLevel() {
+        return level;
```

Comment:
To make it *slightly* harder for people to create concurrency bugs by using the level inside the additional section renderer (instead of the proper snapshot region), maybe this getter should throw if `Minecraft.getInstance().isSameThread()` returns false. Just a suggestion, doesn't need to block merging.